### PR TITLE
Add Membership.get_staff()

### PR DESCRIPTION
**Pull request by pwilkins**
Implement a get_staff() function in Membership to retrieve a course's staff roster.  
Ready for review.

Addresses issue #34 Add get_staff() function
@pdpinch @carsongee @ichuang

### CODE REVIEW

File: pylmod/membership.py [hunk 1]
```diff
@@ -20,5 +20,142 @@ def __init__(
...
+                                u'nickName': u'Benjamin',
+                                u'personId': 333333,
+                                u'sortableName': u'Franklin, Benjamin',
+                                u'surname': u'Franklin',
+                                u'year': None}]},
+                    u'message': u'',
+                    u'status': 1
+                }
+                {
+                    u'recitation':
```
Comment (carsongee) [inline pylmod/membership.py]: What is this madness?

File: pylmod/membership.py [hunk 2]
```diff
@@ -20,5 +20,142 @@ def __init__(
...
+            list: list of dictionaries containing staff data
+
+            An example return value is:
+
+            .. code-block:: python
+
+                {
+                    u'data':
+                    {
+                        u'COURSE_ADMIN':
```
Comment (carsongee) [inline pylmod/membership.py]: Because this is structured upside down like sections, we should provide a method to reverse it.  Also a "simple" argument that changes the return to just a list of `accountEmails`

File: pylmod/membership.py (hunk 1 above)
Comment (pwilkins) [inline pylmod/membership.py]: April Fool?  Actually a cut/paste error.

**Comment by pwilkins** [comment]
Ready for review. @carsongee @pdpinch

**Comment by carsongee** [comment]
Looks like there is new code that doesn't have coverage, and not all your test code is being run either.

File: pylmod/membership.py [hunk 3]
```diff
@@ -9,7 +10,7 @@
 
 class Membership(Base):  # pylint: disable=too-few-public-methods
     """Provide API for functions that return group membership data from
-    MIT Learning Modules service.
+    MIT Learning Modules web service.
```
Comment (carsongee) [inline pylmod/membership.py]: nit: Web

File: pylmod/membership.py [hunk 4]
```diff
@@ -18,7 +19,106 @@ def __init__(
...
+                courseId=course_id or self.course_id
+            ),
+            params=None
+        )
+
+        if simple:
+            return staff_data
+        return staff_data
+
+    def get_course_id(self, course_uuid=''):
```
Comment (carsongee) [inline pylmod/membership.py]: nit: I think it makes more sense to have this next to `__init__` since it is called there.

File: pylmod/gradebook.py [hunk 5]
```diff
@@ -82,6 +82,64 @@ def get_gradebook_id(self, gbuuid):
             raise PyLmodUnexpectedData(failure_messsage)
         return gradebook['data']['gradebookId']
 
+    def get_options(self, gradebook_id):
```
Comment (carsongee) [inline pylmod/gradebook.py]: This seems like a a different logical change than what this PR is doing as there are no dependencies either way from this to get_staff or get_staff to this.  Open this as another PR?

File: pylmod/membership.py [hunk 6]
```diff
@@ -18,7 +19,106 @@ def __init__(
...
+        """Get course id based on uuid
+
+        Args:
+            uuid (str): course uuid, i.e. /project/mitxdemosite
+
+        Returns:
+            str: numeric course id
```
Comment (carsongee) [inline pylmod/membership.py]: why are we returning this as a string if it is a number?

File: pylmod/membership.py [hunk 7]
```diff
@@ -18,7 +19,106 @@ def __init__(
...
+                }
+    """
+        staff_data = self.get(
+            'courseguide/course/{courseId}/staff'.format(
+                courseId=course_id or self.course_id
+            ),
+            params=None
+        )
+
+        if simple:
```
Comment (carsongee) [inline pylmod/membership.py]: simple is returning the same thing as not simple, why is that?

File: pylmod/membership.py [hunk 8]
```diff
@@ -18,7 +19,106 @@ def __init__(
...
+            list: list of dictionaries containing staff data
+
+            An example return value is:
+
+            .. code-block:: python
+
+                {
+                    u'response':
+                    {
+                        u'docs':
```
Comment (carsongee) [inline pylmod/membership.py]: what is this madness?  Also, I would think simple would just return the displayNames per the other `simple` conventions.  Also, is there ever any other key besides `docs`?  If not, we should return the contents of docs, since the other data is meta data, if not, we should probably do an unravel and set an attribute with each staff entry that is whatever data "docs" is supposed to represent so the structure is flat.

File: pylmod/tests/test_membership.py [hunk 9]
```diff
@@ -2,13 +2,151 @@
...
+import httpretty
+import json
+
+
 from pylmod import Membership
 from pylmod.tests.common import BaseTest
 
 
-class TestGradebook(BaseTest):
+class TestMembership(BaseTest):
```
Comment (carsongee) [inline pylmod/tests/test_membership.py]: whoops :flushed:

File: pylmod/tests/test_membership.py [hunk 10]
```diff
@@ -2,13 +2,151 @@
...
+                u'params':
+                {
+                    u'wt': u'json'
+                },
+                u'qtime': 0,
+                u'status': 0
+            }
+    }
+
+    def _register_get_membership(self, send_data=True):
```
Comment (carsongee) [inline pylmod/tests/test_membership.py]: this isn't being called in tests, does this even need to be here? Also, it has a print statement in it.

File: pylmod/base.py [hunk 11]
```diff
@@ -26,9 +26,12 @@ class Base(object):
         urlbase (str): The URL of the LMod web service. i.e.
             ``learning-modules.mit.edu`` or ``learning-modules-test.mit.edu``
     """
-    #: The default Stellar site
+    #: The default Stellar gradebook uuid
     GBUUID = 'STELLAR:/project/mitxdemosite'
```
Comment (carsongee) [inline pylmod/base.py]: Are either of these ever even used?  It doesn't seem like it.

**Comment by carsongee** [comment]
This is really unfortunate data that the API is returning, is the gradebook side better?  Weren't we planning on doing a PR for that as well, or did we just decide one was good enough?  I don't fully recall, but if the gradebook side returns actual e-mails or user IDs or something, we will definitely want to add that at some point (though it would be out of scope for this PR).

File: pylmod/tests/test_membership.py (hunk 10 above)
Comment (pwilkins) [inline pylmod/tests/test_membership.py]: About the print statement: I found one in the GradeBook tests, so I assumed it was the thing to do.  I should take them both out.

**Comment by pwilkins** [comment]
About that Membership data: The comprehensive staff data you saw previously was returned by the Gradebook API.  I have opened a ticket in to LMod why these are so different.

**Comment by pwilkins** [comment]
I would have expected that the local coverage test would have reported this lack of coverage.  It doesn’t.  :-(

**Comment by carsongee** [comment]
For local coverage, sometimes you need to delete your .coverage file.

**Comment by pwilkins** [comment]
this PR is ready for another review.  @carsongee @pdpinch

File: pylmod/membership.py [hunk 12]
```diff
@@ -18,7 +19,89 @@ def __init__(
...
+            ),
+            params=None
+        )
+        if 'response' not in course_data:
+            failure_messsage = ('Error in get_course_id '
+                                'for {0} - no response'.format(course_data))
+            log.error(failure_messsage)
+            raise PyLmodUnexpectedData(failure_messsage)
+
+        return course_data['response']['docs'][0]['id']
```
Comment (carsongee) [inline pylmod/membership.py]: This is heavily blind indexing, I think you probably want to catch the exception here and raise a PyLmod exception for the entire object address.

**Comment by pwilkins** [comment]
This PR is again ready for review following the last round of review comments.  @carsongee @pdpinch

File: pylmod/membership.py [hunk 13]
```diff
@@ -18,7 +19,90 @@ def __init__(
...
+        """
+        course_data = self.get(
+            'courseguide/course?uuid={uuid}'.format(
+                uuid=course_uuid or self.course_id
+            ),
+            params=None
+        )
+        try:
+            return course_data['response']['docs'][0]['id']
+        except Exception:
```
Comment (carsongee) [inline pylmod/membership.py]: Weird that this is passing tests, I guess flakes doesn't check for broad exceptions.  This should only except on specific errors, i.e. `TypeError` when the data isn't a dictionary and `KeyError` when it is a dictionary but the key doesn't exist.  You will probably want a test that exercises both.

File: pylmod/membership.py [hunk 14]
```diff
@@ -18,7 +19,90 @@ def __init__(
...
+            failure_message = ('Error in get_course_id '
+                               'for {0} - no response'.format(course_data))
```
Comment (carsongee) [inline pylmod/membership.py]: This message isn't really quite as descriptive any more.  I would do `log.exception('Error in get_course_id - got %r', course_data)`

File: pylmod/tests/test_membership.py [hunk 15]
```diff
@@ -2,23 +2,172 @@
...
+    def test_get_staff(self):
+        """Verify that we can get staff roster as expected."""
+        self._register_get_staff()
+        test_membership = Membership(self.CERT, self.URLBASE)
+        staff_list = test_membership.get_staff(course_id=self.COURSE_ID)
+        self.maxDiff = None
+        self.assertEqual(staff_list, self.STAFF_BODY['response']['docs'])
+
+    @httpretty.activate
+    def test_get_course_id(self):
```
Comment (carsongee) [inline pylmod/tests/test_membership.py]: nit: it is kind of nice if tests order with their methods, but seriously this doesn't need to be fixed

File: pylmod/tests/test_membership.py [hunk 16]
```diff
@@ -2,23 +2,172 @@
...
+        """Verify that we can the course id by providing the course uuid."""
+        self._register_get_course_id()
+        test_base = Membership(self.CERT, self.URLBASE)
+        course_id = test_base.get_course_id(self.CUUID)
+        self.assertEqual(course_id, self.COURSE_ID)
+
+        # Remove data and assert exception raised
+        self._register_get_course_id(False)
+        with self.assertRaises(PyLmodUnexpectedData):
```
Comment (carsongee) [inline pylmod/tests/test_membership.py]: Paired with my first suggestion, I would mock the log and raise first a key error and second a type error and assert the correct log message is given for each.

**Comment by pwilkins** [comment]
This should be ready to merge.  I made the last correction to the code and rebased. @pdpinch @carsongee

File: pylmod/tests/test_membership.py [hunk 17]
```diff
@@ -2,23 +2,161 @@
...
+        )
+
     def test_constructor(self):
         """Verify constructor does as expected i.e. append URLBASE with
         ``service/membership/``
         """
         # Strip off base URL to make sure it comes back
         urlbase = self.URLBASE[:-1]
-        test_membership = Membership(self.CERT, urlbase)
+        test_base = Membership(self.CERT, cuuid=None, urlbase=urlbase)
```
Comment (carsongee) [inline pylmod/tests/test_membership.py]: Why did you change these?

Comment (pwilkins) [inline pylmod/tests/test_membership.py]: you mean change the variable name or change the parameter list?  I didn't think that the variable name was significant and I thought that the parameter list was equivalent only more explicit.

Comment (carsongee) [inline pylmod/tests/test_membership.py]: The variable name since it is an instance of the Membership class, not the Base class.  Also, just curious, is this new and I didn't notice, or did you recently change the variable name?

**Comment by carsongee** [comment]
:gem: :beers:
